### ISSUE TITLE
[expr] Clarify which syntax components designate expressions

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/CWG#76]
Full name of submitter (unless configured in github; will be published with the issue):  Jim X

In the current draft, we intend to use the syntax notation *X-expression* to designate an expression. However, we didn't have a clear wording to clarify this intent.  There is only one note in [expr] that implies this point. 

[expr.pre] p1
> [Note 1: **[expr]** defines the **syntax**, order of evaluation, and meaning **of expressions**. An expression is a sequence of operators and operands that specifies a computation. An expression can result in a value and can cause side effects. — end note]

which implies that all syntax components defined in the whole [expr] are expressions. Such as *id-expression*, *assignment-expression*, and so on. As well, [gram.expr] also lists all such syntax notations that are expressions.  

However,  an *id-expression* that is an expression can be an *identifier*, and a type such as *class-name* can also be an identifier, When an *identifier*(lower-level syntax) appears in grammar, we should have an explicit way to distinguish whether the syntax is considered to be an expression or something else. This should be distinguishable by analyzing the higher-level syntax to which the lower-level syntax belongs. 

For example:
````cpp
  T{0};
  int a = U;

  void operator "" _show(const char*);
  auto ptr = "";
````
The *identifier* `T` belongs to *class-name* that belongs to *type-name* that belongs to *simple-type-specifier*, whereas the *identifier* `U` belongs to *id-expression* that belongs to ... that belongs to *assignment-expression* that belongs to *initializer-clause* that belongs to *initializer*. As the lower-level syntax, both `T` and `U` are *identifiers*, however, the former is a type while the latter is an expression. That is true for *string-literal* in this example. 

I think it is exactly the intent of defining these higher-level syntaxes. However, we lack the relevant form wording to explicitly convey these meanings. 

### Suggested resolution
Augment [syntax] p2
> *X-expression* is an expression with no context-dependent meaning. 

Change *declarator-id* to
> declarator-id:
>> - ...<sub>opt</sub> unqualified-id
>> - ...<sub>opt</sub> qualified-id

After this modification, it is clear that any defined grammar that comprises the syntax such as *X-expression* designates that syntax is interpreted as an expression. 

````cpp
int a = T;
````
The declaration matches grammar *decl-specifier-seq declarator initializer* where the *initializer* matches *= assignment-expression*, that is the *identifier* `T` in this declaration is interpreted as an expression.

**Comment by jensmaurer** [comment]
Class member access [expr.ref] is another situation where we use _id-expression_, but it's not a standalone expression (it's a designator for a class member, instead).

I think we should introduce a separate grammar production, e.g. _qual-unqual-id_, for _declarator-id_ and class member access, and reserve _id-expression_ for the expression uses.

**Comment by xmh0511** [comment]
> I think we should introduce a separate grammar production, e.g. qual-unqual-id, for use in declarator-id and class member access, and reserve id-expression for the expression uses. 

Agree. That will make the semantics clear. Users will clearly know which are expressions, which are types, and which are something else in a particular grammar by using the syntaxes with the clear meaning.
